### DISCUSSION TITLE
napalm-panos not working

### DISCUSSION THREAD

**Discussion by misho23-23** [netbox-community/netbox#12331] [Q&A]
Hello,

I'm trying to intergate <code>napalm_panos</code> with <code>netbox</code> but I'm receiving this error message 

<strong>Missing required packages</strong> - This installation of NetBox might be missing one or more required Python packages. These packages are listed in <code>requirements.txt</code> and <code>local_requirements.txt</code>, and are normally installed as part of the installation or upgrade process. To verify installed packages, run <code>pip freeze</code> from the console and compare the output to the list of required packages.

<strong>WSGI service not restarted after upgrade</strong> - If this installation has recently been upgraded, check that the WSGI service (e.g. gunicorn or uWSGI) has been restarted. This ensures that the new code is running.

The complete exception is provided below:
<strong>&lt;class &#x27;ModuleNotFoundError&#x27;&gt;</strong>
```
No module named napalm_base

Python version: 3.10.11
NetBox version: 3.4.8
``` 

I tried to install it manually this package <code>napalm_base</code> via pip
```
 pip install napalm_base
``` 
![image](https://user-images.githubusercontent.com/130065618/233966083-bb9e73fc-1e14-4092-9b97-a53ddd52ac60.png)

Environment
```
netbox]# which pip
/usr/local/bin/pip

netbox]# python -m ensurepip --default-pip

Looking in links: /tmp/tmpm067yy74
Requirement already satisfied: setuptools in /usr/local/lib/python3.10/site-packages (65.5.0)
Requirement already satisfied: pip in /usr/local/lib/python3.10/site-packages (23.1.1)
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
``` 
```
pip freeze | grep napalm
napalm==3.4.1
napalm-panos==0.6.0
``` 
Simulation:
Devices -> Device_ID -> Status OR LLDP Neighbors OR Config

OR
Rest API call with Napalm method doesn't work with NetBox - it's same device ID
```
http://<NetBox IP>/api/dcim/devices/5/napalm/?method=get_facts
``` 
with same error message

Otherwise, NAPALM with panos driver works fine
``` 
#napalm --vendor panos --user <username> --password <psw> <IP> call get_facts
``` 
this command works fine via CLI and I receive output
``` 
{
    "hostname": "FW-XXXX-01",
    "vendor": "Palo Alto Networks",
    "uptime": 9514829,
    "os_version": "10.1.8",
    "serial_number": "XXXXXXXXXXXX",
    "model": "PA-450",
    "fqdn": "N/A",
    "interface_list": [
        "ethernet1/1",
        "ethernet1/2",
        "ethernet1/2.70",
        "ethernet1/3",
        "ethernet1/3.40",
        "ethernet1/5",
        "ethernet1/5.50",
        "loopback",
        "loopback.1",
        "tunnel",
        "tunnel.1",
        "tunnel.2",
        "tunnel.3",
        "tunnel.4",
        "vlan"
    ]
}
```

**Comment by candlerb** [comment]
> I tried to install it manually this package `napalm_base` via pip

That won't work unless you've first activated the virtualenv (Netbox only sees packages which are installed in its venv)

Recommended approach is to create `local_requirements.txt` listing the packages(s) you want to install, as documented [here](https://docs.netbox.dev/en/stable/installation/3-netbox/#napalm).  Then run `./upgrade.sh` to reinstall the netbox virtualenv.  This approach ensures you'll always have the extra packages you requested, every time you upgrade netbox.

**Reply by misho23-23** [reply]
> > I tried to install it manually this package `napalm_base` via pip
> 
> That won't work unless you've first activated the virtualenv (Netbox only sees packages which are installed in its venv)
> 
> Recommended approach is to create `local_requirements.txt` listing the packages(s) you want to install, as documented [here](https://docs.netbox.dev/en/stable/installation/3-netbox/#napalm). Then run `./upgrade.sh` to reinstall the netbox virtualenv. This approach ensures you'll always have the extra packages you requested, every time you upgrade netbox.

thank you for your reply, anyway, I tried to do it via <code>local_requirements.txt</code> and virtual environment <code>vent</code> as well
```
napalm
napalm-panos
napalm-base
``` 
but still receiving error related to <code>pip</code>
![image](https://user-images.githubusercontent.com/130065618/233988363-c0a2dce2-614d-4096-b4fb-dda07ec92ea6.png)
in this case via <code>local_requirements.txt</code> it's 
<code>ModuleNotFoundError: No module named 'pip.req'</code>

**Reply by candlerb** [reply]
Looks like brokenness in napalm itself - if you can reproduce this without Netbox then raise it with napalm.

I found a similar issue for napalm-fortios:
https://github.com/napalm-automation-community/napalm-fortios/issues/57#issuecomment-1445146465

**Reply by misho23-23** [reply]
> Looks like brokenness in napalm itself - if you can reproduce this without Netbox then raise it with napalm.
> 
> I found a similar issue for napalm-fortios: [napalm-automation-community/napalm-fortios#57 (comment)](https://github.com/napalm-automation-community/napalm-fortios/issues/57#issuecomment-1445146465)

I tried to use also PYTHON=/usr/bin/python3.8 but still problem...
unfortunately, this is only in Netbox, NAPALM with panos driver works fine
```
#napalm --vendor panos --user <username> --password <psw> <IP> call get_facts
``` 
this command works fine via CLI and I receive output
``` 
{
    "hostname": "FW-XXXX-01",
    "vendor": "Palo Alto Networks",
    "uptime": 9514829,
    "os_version": "10.1.8",
    "serial_number": "XXXXXXXXXXXX",
    "model": "PA-450",
    "fqdn": "N/A",
    "interface_list": [
        "ethernet1/1",
        "ethernet1/2",
        "ethernet1/2.70",
        "ethernet1/3",
        "ethernet1/3.40",
        "ethernet1/5",
        "ethernet1/5.50",
        "loopback",
        "loopback.1",
        "tunnel",
        "tunnel.1",
        "tunnel.2",
        "tunnel.3",
        "tunnel.4",
        "vlan"
    ]
}
```
